Add --config cli option for overwriting the path to config.py

xaverdh opened the pull request:
This adds a new cli flag for setting the path to config.py.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4286)
<!-- Reviewable:end -->

*Edit by @The-Compiler to auto-close the issue: Fixes #4132*

The-Compiler commented on qutebrowser/qutebrowser.py:
```diff
@@ -63,6 +63,7 @@ def get_argparser():
                                      description=qutebrowser.__description__)
     parser.add_argument('-B', '--basedir', help="Base directory for all "
                         "storage.")
+    parser.add_argument('-C', '--config', help="Path to config.py")
```
I'd prefer to call it `--config-py` so it's clear that it's only a path to a `config.py` file, not a config folder.

xaverdh commented:
Ok, I renamed the option into --config-py and fixed the tests.
About adding a new test, I can try, but this will be trial and error very much (never worked with pytest and the like). Also I can't run the tests locally atm, due to having a slightly older pyqt version.

The-Compiler commented on tests/unit/config/test_configinit.py:
```diff
@@ -65,6 +67,22 @@ def configdata_init(monkeypatch):
 class TestEarlyInit:
 
+    def test_config_py_path(self, args):
+        with tempfile.NamedTemporaryFile(encoding = 'utf-8' ,
```
Instead of the `tempfile` module, please use the [`tmpdir` fixture](https://docs.pytest.org/en/latest/tmpdir.html) provided by pytest. The object you get is a [py.path.local](https://py.readthedocs.io/en/latest/path.html) instance, and there's also some [docs about fixtures](https://docs.pytest.org/en/latest/fixture.html).

The-Compiler commented on tests/unit/config/test_configinit.py:
```diff
@@ -65,6 +67,22 @@ def configdata_init(monkeypatch):
+            args.config_py = conf_file.name
+            configinit.early_init(args)
+
+            # Make sure things have been init'ed
```
I don't think those two tests are needed here.

The-Compiler commented on tests/unit/config/test_configinit.py:
```diff
@@ -65,6 +67,22 @@ def configdata_init(monkeypatch):
 
 class TestEarlyInit:
 
+    def test_config_py_path(self, args):
```
You'll need to use the `init_patch` fixture here (like `test_config_py` does) so stuff which is in the way gets patched away.

jgkamat commented:
@xaverdh, do you have time to keep working on this? There are still some comments un-resolved.

xaverdh commented:
> @xaverdh, do you have time to keep working on this? There are still some comments un-resolved.
> 
> This seems pretty simple, and I'd like to see it added, this would be very nice to have during testing :]
> 
> The implementation itself seems pretty solid, although I would have prefered `or` instead of the ternary (intentionally ignoring if it's empty string), but I'm fine either way.

I am stuck on adding tests. I never worked with any test framework in python (I come from haskell land mostly) , and don't have enough time to work on this atm.
If you want to take this over, feel free to do so :-)

The-Compiler commented on tests/unit/config/test_configinit.py:
```diff
@@ -65,6 +66,20 @@ def configdata_init(monkeypatch):
+        args.config_py = str(f)
+        configinit.early_init(args)
+
+        # Make sure things have been init'ed
```
@jgkamat Not sure if you've seen it already:

https://github.com/qutebrowser/qutebrowser/pull/4286#discussion_r222755876
> I don't think those two tests are needed here.

jgkamat commented:
Florian Bruhin writes:

> The-Compiler requested changes on this pull request.
>
> @jgkamat what's your take on this?
>
>> Like you said, `:config-write-py` uses the default config path right now, and so do `:config-edit` and `:config-source`. Maybe they should also use the modified path? Not sure about it though.

I personally am torn between where :config-edit and :config-write-py should
go. I don't think they should go to the overriden path so they apply on next
launch, but that behavior is also confusing as it mixes the logic a bit. I do
think :config-source should definitely go to the overridden file though.

Either way, we'll need some function or variable in standarddir or config to
hold the overriden config.py location. Where do you think it should go?

The-Compiler commented:
I think I'd prefer all three to go to the same location (i.e. what's specified via `--config-py`) for consistency. If `:config-source` sources the overridden config file, I feel like `:config-edit` should edit the same file as well. And at that point, I guess it's best to make `:config-write-py` go there as well.

As for where to save the location: I don't really have a preference there. I guess it'd fit best as a global in `configfiles.py` or maybe `standarddir.py`.

The-Compiler commented:
@jgkamat Should this still have a "status: WIP" label?

jgkamat commented:
Florian Bruhin writes:

> @jgkamat Should this still have a "status: WIP" label?

I think I added that due to missing tests. I think I should have enough 
coverage now so I'll remove it, but let me know if I'm missing anything

The-Compiler commented:
Oh, looks like tests are failing: [py35-pyqt571](https://travis-ci.org/qutebrowser/qutebrowser/jobs/522048955) (probably needs some additional `str()` for `tmpdir` objects), [py37-pyqt512](https://travis-ci.org/qutebrowser/qutebrowser/jobs/522048960) - @jgkamat do you have time take a look somewhen this week? Otherwise I can pick this up.

The-Compiler commented on tests/helpers/fixtures.py:
```diff
@@ -545,9 +545,22 @@ def config_tmpdir(monkeypatch, tmpdir):
 
+@pytest.fixture
+def config_py_arg(tmpdir, monkeypatch):
+    """Arguments needed for the config to init."""
```
Huh? This doesn't seem to provide any arguments?

The-Compiler commented on tests/helpers/fixtures.py:
```diff
@@ -545,9 +545,22 @@ def config_tmpdir(monkeypatch, tmpdir):
+@pytest.fixture
+def config_py_arg(tmpdir, monkeypatch):
+    """Arguments needed for the config to init."""
+    f = tmpdir.join('temp_config.py')
```
```suggestion
    f = tmpdir / 'temp_config.py'
```

for consistency.

The-Compiler commented on tests/unit/config/test_configcommands.py:
```diff
@@ -493,6 +493,14 @@ def test_config_source(self, tmpdir, commands, config_stub, config_tmpdir,
 
+    def test_config_py_arg_source(self, commands, config_py_arg, config_stub):
+        assert config_stub.val.content.javascript.enabled
+        config_stub.val.search.ignore_case = 'always'
```
Is this needed?

The-Compiler commented on tests/unit/utils/test_standarddir.py:
```diff
@@ -253,6 +253,24 @@ def test_basedir_relative(self, tmpdir):
             assert standarddir.config() == str(basedir / 'config')
 
+    def test_config_py_arg(self, tmpdir):
+        basedir = (tmpdir / 'basedir')
```
```suggestion
        basedir = tmpdir / 'basedir'
```

The-Compiler commented on tests/unit/utils/test_standarddir.py:
```diff
@@ -253,6 +253,24 @@ def test_basedir_relative(self, tmpdir):
+            assert standarddir.config_py() == str(basedir / 'config.py')
+
+    def test_config_py_no_arg(self, tmpdir):
+        basedir = (tmpdir / 'basedir')
```
```suggestion
        basedir = tmpdir / 'basedir'
```

The-Compiler commented:
Thanks @xaverdh and @jgkamat!
